### ISSUE TITLE
Reinforcement Learning

### ISSUE THREAD

**Issue by srini1948** [cbovar/ConvNetSharp#55]
RL has been added to original ConvNetJS.
Will you be adding that too?
Any plans for LSTM?

Thanks

**Comment by MarcoMeter** [comment]
+1

Shouldn't be that tough to implement DQN. Maybe I can contribute that in like 8 weeks. Though, I haven't checked out if ConvNetSharp is suitable for my implementation in Unity performance wise.

**Comment by srini1948** [comment]
Thanks – the only reason I ask is because I am a lot more comfortable with C#.

I find the Python array methods extremely confusing – right now I am experimenting with CNTK.

I wish Microsoft will release their C# interface for CNTK – currently they support only eval mode.



Sent from Mail for Windows 10

**Comment by cbovar** [comment]
For DQN you can check out [this repo](https://github.com/dubezOniner/Deep-QLearning-Demo-csharp). It should be easy to adapt it to newer version of ConvNetSharp.

I have worked on LSTM. I will eventually release a 'shakespeare' demo. I only worked on the GPU versions.

**Comment by cbovar** [comment]
I also see a DQN using WPF for display [here too](https://github.com/jankrib/ConvNetSharp/tree/deep-q-demo)

**Comment by MarcoMeter** [comment]
I worked with Deep-QLearning-Demo the past weeks. But it lacks in performance (due to single threaded) and that code is hard to read and to maintain. But well, this was almost completely adapted from the ConvNetJS version, which uses that strange coding convention.

**Comment by srini1948** [comment]
Thanks. I will

**Comment by srini1948** [comment]
Thanks.,

**Comment by srini1948** [comment]
I tried it – very nicely written and the graphics are great!

Sent from Mail for Windows 10

**Comment by MarcoMeter** [comment]
I'm applying DQN to my game BRO ( https://www.youtube.com/watch?v=_mZaGTGn96Y ) right now.
Within the next months , I'll release BRO open source on Github. BRO features an AI framework and a match sequence editor for match automation. The game is done with Unity.

Right now I need a much faster DQN implementation. The DQN Demo from above lacks in that, so the training time takes 30 minutes. That's why I'm considering to contribute DQN to this repo.

And this is a video about the AI framework and the match sequence editor https://www.youtube.com/watch?v=EE7EqoaOL34

**Comment by srini1948** [comment]
If ConvNetSharp works with GPU can the DeepQLearning Demo be modified to use the Volume.GPU structure? It will be much faster.

Sent from Mail for Windows 10

**Comment by MarcoMeter** [comment]
Hey,

if anybody has some ideas for testing the **to be** implemented DQN algorithm, please let me know.

So far I've got these ideas for integration testing:

- ContNetJs's apple and poison example (windows forms), just like the already mentioned C# port (Deep-QLearning-Demo-csharp)
- slot machine (just console application)
- a moving target, which has to be shot by the agent (maybe Unity game engine)
- agent has to move a basket to catch fruits and to avoid stuff like poison (maybe Unity game engine)

I'll probably find more through research.

After that, I'll start with the DQN implementation. I'll probably start with the implementation of "Deep-QLearning-Demo-csharp". Then I'll compare it to the Python implementation done by DeepMind for the Atari games.

**Comment by srini1948** [comment]
Thanks.
Srini


Sent from my Galaxy Tab A
-------- Original message --------From: MarcoMeter <notifications@github.com> Date: 7/24/17  12:23 PM  (GMT-05:00) To: cbovar/ConvNetSharp <ConvNetSharp@noreply.github.com> Cc: srini1948 <ranjisrini@gmail.com>, Author <author@noreply.github.com> Subject: Re: [cbovar/ConvNetSharp] Reinforcement Learning (#55) 
Hey,
if anybody has some ideas for testing the to be implemented DQN algorithm, please let me know.
So far I've got these ideas for integration testing:

ContNetJs's apple and poison example (windows forms), just like the already mentioned C# port (Deep-QLearning-Demo-csharp)
slot machine (just console application)
a moving target, which has to be shot by the agent (maybe Unity game engine)
agent has to move a basket to catch fruits and to avoid stuff like poison (maybe Unity game engine)

I'll probably find more through research.
After that, I'll start with the DQN implementation. I'll probably start with the implementation of "Deep-QLearning-Demo-csharp". Then I'll compare it to the Python implementation done by DeepMind for the Atari games.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/cbovar/ConvNetSharp","title":"cbovar/ConvNetSharp","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/cbovar/ConvNetSharp"}},"updates":{"snippets":[{"icon":"PERSON","message":"@MarcoMeter in #55: Hey,\r\n\r\nif anybody has some ideas for testing the **to be** implemented DQN algorithm, please let me know.\r\n\r\nSo far I've got these ideas for integration testing:\r\n\r\n- ContNetJs's apple and poison example (windows forms), just like the already mentioned C# port (Deep-QLearning-Demo-csharp)\r\n- slot machine (just console application)\r\n- a moving target, which has to be shot by the agent (maybe Unity game engine)\r\n- agent has to move a basket to catch fruits and to avoid stuff like poison (maybe Unity game engine)\r\n\r\nI'll probably find more through research.\r\n\r\nAfter that, I'll start with the DQN implementation. I'll probably start with the implementation of \"Deep-QLearning-Demo-csharp\". Then I'll compare it to the Python implementation done by DeepMind for the Atari games."}],"action":{"name":"View Issue","url":"https://github.com/cbovar/ConvNetSharp/issues/55#issuecomment-317476274"}}}

**Comment by cbovar** [comment]
Maybe you could also try on a very simple task, reproduce the input:
* 0 -> 0
* 1 -> 1

It may fit in an unit test.

**Comment by MarcoMeter** [comment]
I wrote a simple slot machine (console app) using 3 reels. Just hold down space to start the slot machine and to stop each reel one by one.

New items for the reels' slots are sampled from a specific probability distribution.

In the end, the agent has to decide when to stop the first reel, the second reel and finally the third one.
(I should consider to let the AI decide which reel to stop first to have a few more dimensions on the outputs)

[SlotMachine.zip](https://github.com/cbovar/ConvNetSharp/files/1172337/SlotMachine.zip)

Given this slot machine example, I'm going to approach now the DQN implementation.

**Comment by srini1948** [comment]
FYI:

As you probably know the Arcade Learning Environment ( containing simulation of Atari games) is available from OpenAI for Linux and MacOS.

I managed to compile it on Windows from source to create a DLL that can be used with Python.

I was wondering if this will be useful for c# too.


Srini


Sent from Mail for Windows 10

**Comment by MarcoMeter** [comment]
Creating an interface between Python and C# might end up consuming too much time. I know there is the so called IronPython (http://ironpython.net/) Library which allows to use Python in C#, but I haven't really looked into it.

**Comment by srini1948** [comment]
I was suggesting using the DLL directly from C#.

See https://gym.openai.com/

It exposes the interface in the attached header file that allows you to load any of the Atari games such as Pong. During training episodes this AI will update status.

Sent from Mail for Windows 10

**Comment by MarcoMeter** [comment]
Here is an update on the progress referring to a commit on the DQN branch of my fork:

Added major chunks of the DeepQLearner.cs [WIP]
A few TODOs left before testing and verification:
- TODO: Overload or modify RetrievePolicy() to make use of Volumes, return output Volume from the net as well
- TODO: Overload or modify GetNetInput() to make use of Volumes
- TODO: Compute loss
- TODO: Verify the consistency of the composed neural net upon initializing the DeepQLearner

https://github.com/MarcoMeter/ConvNetSharp/commit/5711468362d6f3551f82bad1e24d784e31f59a4b

**Comment by MarcoMeter** [comment]
And there is one more major thing on the list:

Adding a regression layer. I guess there is no regression layer implemented yet, right?

**Comment by cbovar** [comment]
It seems that **RegressionLayer** disappeared at some point (from tag 0.3.2). I will try to reintroduce it this week end.

**Comment by MarcoMeter** [comment]
Maybe this is related to this commit, because the file 'F:\Repositories\ConvNetSharp\src\ConvNetSharp\Layers\RegressionLayer.cs' got removed:

Commit: 56fec45931434ab1e39f869495eebb21941218ee [56fec45]
Parents: 5a47e2e2a7, 37cdfbf451
Author: Augustin Juricic <ajuricic@neogov.net>
Date: Dienstag, 28. März 2017 11:18:18
Committer: Augustin Juricic
Merge remote-tracking branch 'github/master' into develop

**Comment by cbovar** [comment]
I think I have never implemented _RegressionLayer_ since ConvNetSharp handles batch.

**Comment by cbovar** [comment]
_RegressionLayer_ committed

**Comment by MarcoMeter** [comment]
Great, thanks. I'll move on soon.

**Comment by MarcoMeter** [comment]
As of now, I'm struggling with the issue that the computed action values grow exponentially towards positive or negative infinity.

**Comment by cbovar** [comment]
Have you tried with a lower learning rate? E.g. 0.001

**Comment by MarcoMeter** [comment]
The learning rate slightly delays this outcome.

Nevertheless, for the outputs I'm expecting values to be less than 2. Just because of the fact that the maximum reward for the slot machine example is 1, which is probably handed out after making at least 3 decisions.

**Comment by MarcoMeter** [comment]
I'm still trying to figure out the issue. Maybe I'm misusing the volume class, or I might don't have enough experience with the actual implementation of neural nets (like understanding every single detail of the regression layer implementation). So I'm dropping some more information.


Here is some pseudo code [(Matiisen, 2015)](http://neuro.cs.ut.ee/demystifying-deep-reinforcement-learning/) featuring the core pieces of the algorithm:

`initialize replay memory D
initialize action-value function Q with random weights
observe initial state s
repeat
    select an action a
        with probability ε select a random action
        otherwise select a = argmaxa’Q(s,a’)
    carry out action a
    observe reward r and new state s’
    store experience <s, a, r, s’> in replay memory D

    sample random transitions <ss, aa, rr, ss’> from replay memory D
    calculate target for each minibatch transition
        if ss’ is terminal state then tt = rr
        otherwise tt = rr + γmaxa’Q(ss’, aa’)
    train the Q network using (tt - Q(ss, aa))^2 as loss

    s = s'
until terminated`

And this is the stated loss function for training:

![lossfunction](https://user-images.githubusercontent.com/6996955/28879072-6c8205f0-77a1-11e7-957c-bdf9c60c2b9e.png)

For the DQN implementation of Karpathy this loss function seems to be not present. The regression layer implementation looks to be similar (comparing Karpathy and this repo). For the rest, everything is implemented accordingly (i.e. sampling experiences for computing new q values).


Using the [Deep Q Learning Demo CSharp](https://github.com/dubezOniner/Deep-QLearning-Demo-csharp), the output values for the slot machine stay below 0.02.
[SlotMachine.zip](https://github.com/cbovar/ConvNetSharp/files/1194334/SlotMachine.zip)

**Comment by cbovar** [comment]
I haven't had time to look at the code yet. But you could maybe make the problem even simpler (like [this](https://medium.com/@awjuliani/super-simple-reinforcement-learning-tutorial-part-1-fd544fab149)) to make it easier to debug.

**Comment by MarcoMeter** [comment]
I could implement an example for contextual bandits according to [Bandit Dungeon Demo](https://github.com/Unity-Technologies/BanditDungeon) (Example form the same author of your provided link).

I just fear that the bandit examples are not complex enough for using a policy network. At least it can be observed, if the q-values grow to infinity or not.

**Comment by srini1948** [comment]
Policy networks are a challenge. I have been trying to reproduce Karpathy’s policy gradient solution for Pong using Microsoft CNTK – for some reason the training does not converge even though I am using the same network architecture and parameters.

**Comment by MarcoMeter** [comment]
The only news I have is that I'm working on a [different example](https://github.com/MarcoMeter/Basket-Catch-Deep-Reinforcement-Learning) (made with Unity). This example is about controlling a basket to catch rewarding items and to avoid punishing ones.

![newexmaple](https://user-images.githubusercontent.com/6996955/29168351-c61d79c2-7dce-11e7-8e93-7b141efdf76a.png)

Concerning the DQN implementation I'm still stuck. I hope that Cedric can find some time to check the usage of Volumes.

**Comment by cbovar** [comment]
Sorry guys. I have been very busy with my new job. I'll try to look at this soon.

**Comment by MarcoMeter** [comment]
I just tested the implementation on the apples & poison example. The issue of exploding output values is observed as well.

I didn't add the example to the version control, since the code is not well written but functional (I took the known implementation and just substituted the DQN parts).

[ApplesPoisonDQNDemo.zip](https://github.com/cbovar/ConvNetSharp/files/1217320/ApplesPoisonDQNDemo.zip)

**Comment by MarcoMeter** [comment]
Just some update:

I created a UI for displaying the training progress. The red graph plots the average reward and the blue one the average loss. I resolved a bug concerning the epsilon exploration strategy (epsilon was always equal to 1 due to an integer division).

![gui](https://user-images.githubusercontent.com/6996955/29453824-4bbe2ed6-840b-11e7-83e4-af1526a321be.png)

As Cederic fixed a bug of the regression layer, the outputs do not explode anymore. Regardless, I did not achieve a good behavior for the slot machine yet. Though I came up with a new reward function, which signals rewards based on the result of one stopped reel. The first stop rewards the agent by the item of slot (e.g. 0.5 for a cherry or 1 for a 7). Stopping the second or third reel rewards the agent by 1 for a matching item. For failure, the agent is punished by -0.5. Waiting does not punish or reward the agent. Most of the time the agent learns to wait. It seems to be that this way, any punishments are avoided.

I probably try to focus now on the Apples and Poison demo, because suitable hyperparemeters are already known. One drawback is the performance. The referenced demo performs much better. So I'll have to find the bottleneck.
e bottleneck for that as well.

**Comment by cbovar** [comment]
I think you should focus on getting the correct results first. For the performances, we can look at it later (using batch size > 1 and GPU will help)

**Comment by MarcoMeter** [comment]
Still it surprises me that the Apples and Poison demo is much much slower compared to Deep-QLearning-Demo-csharp.

**Comment by cbovar** [comment]
You probably get **BadImageFormatException** because you are in 32bits. GPU only works in 64bits.

**Comment by MarcoMeter** [comment]
Thanks, this solved the BadImageFormatException.

And now its a CudaException thrown at CudaHostMemoryRegion.cs:25

**Comment by cbovar** [comment]
I have fixed the loss computation in the regression layer.

I think there is an issue [here](https://github.com/MarcoMeter/ConvNetSharp/blob/DQN/src/ConvNetSharp.ReinforcementLearning.Deep/DeepQLearner.cs#L251). You get the output of the **FinalState** and update the reward related to current **Action**. However you should get the output related to the **InitialState**.

In ConvNetJs, it only regress on the current **Action** dimension [here](https://github.com/karpathy/convnetjs/blob/5c653cb15a53c3361a6cf21c7277de941ff05352/build/deepqlearn.js#L253).

You could do something like that:
```c#
// Create desired output volume
var desiredOutputVolume = _trainingOptions.Net.Forward(experience.InitialState).Clone();
desiredOutputVolume.Set(actionPolicy.Action, newActionValue);
```

**Comment by MarcoMeter** [comment]
It looks that you are right on that. I missed out on that detail inside the train function.

**Comment by cbovar** [comment]
As for the exception in GPU (_CudaException thrown at CudaHostMemoryRegion.cs:25_), it turns out it's a multi-threading issue: some volume allocation is done in the workerthread whereas the GPU context was acquired on the main thread.

**Comment by masatoyamada1973** [comment]
desiredOutputVolume.Set(actionPolicy.Action, newActionValue)
↓
desiredOutputVolume.Set(experience.Action, newActionValue)

？

**Comment by MarcoMeter** [comment]
Hey,
I wanted to let you guys know that I stopped working on this concern.
I switched to working with Python and the just released [ML Agents of Unity](https://github.com/Unity-Technologies/ml-agents).

**Comment by GospodinNoob** [comment]
@MarcoMeter Hello. Link (https://github.com/MarcoMeter/Basket-Catch-Deep-Reinforcement-Learning) is broken. Is there an opportunity to download the source code of this Unity implementation (Unity project)? Thanks.

**Comment by MarcoMeter** [comment]
@GospodinNoob 
https://github.com/MarcoMeter/Unity-ML-Environments

**Comment by GospodinNoob** [comment]
@MarcoMeter Thanks

**Comment by GospodinNoob** [comment]
@MarcoMeter  Maybe you have a repo with Unity and your DNQ? I am trying to add it, but still, have some misunderstanding of this system. Of course, if it not hard to you)
